mutation_writer/partition_based_splitting_writer.cc

```diff
@@ -21,38 +21,57 @@
 
 #include "mutation_writer/partition_based_splitting_writer.hh"
 
+#include <seastar/core/coroutine.hh>
+
 namespace mutation_writer {
 
 class partition_based_splitting_mutation_writer {
     struct bucket {
         bucket_writer writer;
         dht::decorated_key last_key;
+        size_t data_size = 0;
+        bool in_partition = false;
     };
 
 private:
     schema_ptr _schema;
     reader_permit _permit;
     reader_consumer _consumer;
+    unsigned _max_buckets;
     std::vector<bucket> _buckets;
     bucket* _current_bucket = nullptr;
 
     future<> write_to_bucket(mutation_fragment&& mf) {
+        _current_bucket->data_size += mf.memory_usage();
         return _current_bucket->writer.consume(std::move(mf));
     }
 
-    bucket* create_bucket_for(const dht::decorated_key& key) {
-        return &_buckets.emplace_back(bucket{bucket_writer(_schema, _permit, _consumer), key});
+    future<bucket*> create_bucket_for(const dht::decorated_key& key) {
+        if (_buckets.size() < _max_buckets) {
+            co_return &_buckets.emplace_back(bucket{bucket_writer(_schema, _permit, _consumer), key});
+        }
+        auto it = std::max_element(_buckets.begin(), _buckets.end(), [] (const bucket& a, const bucket& b) {
+            return a.data_size < b.data_size;
+        });
+        if (it->in_partition) {
+            co_await it->writer.consume(mutation_fragment(*_schema, _permit, partition_end{}));
```

Comment:
I think this should be handled higher in the stack.
e.g `consume(partition_start)` below can ensure that `_current_bucket->in_partition == false`,
if not, than the sstable is considered corrupt, and we should throw, no?
If this is run for scrub, then scrub is supposed to handled that in `on_unexpected_partition_start`

Reply:
Right, this check (and the related flag) is superfluous, as the splitter works in terms of entire partitions. I will remove the check and the flag.